File: source/Irrlicht/CIrrDeviceIPhone.m
// Copyright Redshift Software, Inc. 2008
// This file is part of the "Irrlicht Engine".
// For conditions of distribution and use, see copyright notice in irrlicht.h

#import "CIrrDeviceIPhone.h"

#ifdef _IRR_COMPILE_WITH_IPHONE_DEVICE_

#import <UIKit/UIKit.h>
#import <QuartzCore/CAEAGLLayer.h>
#import <OpenGLES/ES1/gl.h>
#import <OpenGLES/ES1/glext.h>
#import <OpenGLES/EAGL.h>
#import <stdio.h>

@interface IrrIPhoneView : UIView
{
}
@end

@implementation IrrIPhoneView
+ (Class) layerClass
{
	return [CAEAGLLayer class];
}
@end

@interface IrrIPhoneDevice : NSObject <UIApplicationDelegate>
{
	@public
	
	EAGLContext * context;
	IrrIPhoneView * view;
	struct MIrrIPhoneDevice * dev;
}
- (id) initWithDevice: (struct MIrrIPhoneDevice*) device;
- (void) dealloc;
- (void) applicationWillResignActive: (UIApplication *) application;
- (void) applicationDidBecomeActive: (UIApplication *) application;
- (void) applicationWillTerminate: (UIApplication *) application;
- (void) displayCreateInWindow: (UIWindow**) window Width: (int) w Height: (int) h;
- (void) displayInitialize: (EAGLContext**) context_ View: (IrrIPhoneView**) view_;
- (void) displayBegin;
- (void) displayEnd;
@end

@implementation IrrIPhoneDevice
- (id) initWithDevice: (struct MIrrIPhoneDevice*) device
{
	if ( (self = [super init]) )
	{
		context = nil;
		view = nil;
		dev = device;
	}
	return self;
}
- (void) dealloc
{
	[super dealloc];
}
- (void) applicationWillResignActive: (UIApplication *) application
{
	// Pause rendering.
	(*(dev->onWindowActive))(dev,0);
}
- (void) applicationDidBecomeActive: (UIApplication *) application
{
	// Resume after non-active pause of rendering.
	(*(dev->onWindowActive))(dev,1);
}
- (void) applicationWillTerminate: (UIApplication *) application
{
	(*(dev->onTerminate))(dev);
}
- (void) displayCreateInWindow: (UIWindow**) window Width: (int) w Height: (int) h
{
	// Create our view.
	CGRect rect;
	rect.origin.x = 0;
	rect.origin.y = 0;
	rect.size.width = w;
	rect.size.height = h;
    view = [[IrrIPhoneView alloc] initWithFrame: rect];
	view.layer.opaque = YES;
	if (nil != *window)
	{
		[(*window) addSubview: view];
	}
	
	// Create the GL context now, so that the driver initializetion
	// can make OpenGL calls.
	context = [[EAGLContext alloc] initWithAPI: kEAGLRenderingAPIOpenGLES1];
	[EAGLContext setCurrentContext: context];
}
- (void) displayInitialize: (EAGLContext**) context_ View: (IrrIPhoneView**) view_
{
	*context_ = context;
	*view_ = view;
	[context renderbufferStorage: GL_RENDERBUFFER_OES fromDrawable: (CAEAGLLayer*)view.layer];
}
- (void) displayBegin
{
	if (context != [EAGLContext currentContext])
	{
		[EAGLContext setCurrentContext: context];
	}
}
- (void) displayEnd
{
	if (context == [EAGLContext currentContext])
	{
		[context presentRenderbuffer: GL_RENDERBUFFER_OES];
	}
}
@end

void irr_device_iphone_display_create(struct MIrrIPhoneDevice * dev,
	void** window, int w, int h)
{
	[((IrrIPhoneDevice*)dev->DeviceM) displayCreateInWindow: (UIWindow**)window Width: w Height: h];
}

void irr_device_iphone_display_init(struct MIrrIPhoneDevice * dev,
	void** context, void** view)
{
	[((IrrIPhoneDevice*)dev->DeviceM) displayInitialize: (EAGLContext**)context View: (IrrIPhoneView**)view];
}

void irr_device_iphone_display_begin(struct MIrrIPhoneDevice * dev)
{
	[((IrrIPhoneDevice*)dev->DeviceM) displayBegin];
}

void irr_device_iphone_display_end(struct MIrrIPhoneDevice * dev)
{
	[((IrrIPhoneDevice*)dev->DeviceM) displayEnd];
}

void irr_device_iphone_create(struct MIrrIPhoneDevice * dev)
{
	dev->DeviceM = [[IrrIPhoneDevice alloc] initWithDevice: dev];
	dev->displayCreate = &irr_device_iphone_display_create;
	dev->displayInit = &irr_device_iphone_display_init;
	dev->displayBegin = &irr_device_iphone_display_begin;
	dev->displayEnd = &irr_device_iphone_display_end;
}

#endif // _IRR_COMPILE_WITH_IPHONE_DEVICE_

File: source/Irrlicht/CIrrDeviceIPhone.cpp
// Copyright (C) 2002-2008 Nikolaus Gebhardt
// Copyright (C) 2008 Redshift Software, Inc.
// This file is part of the "Irrlicht Engine".
// For conditions of distribution and use, see copyright notice in irrlicht.h

#include "CIrrDeviceIPhone.h"

#ifdef _IRR_COMPILE_WITH_IPHONE_DEVICE_

#include <stdio.h>
#include <stdlib.h>
#include <sys/utsname.h>
#include <time.h>
#include "IEventReceiver.h"
#include "os.h"
#include "CTimer.h"
#include "irrString.h"
#include "Keycodes.h"
#include "COSOperator.h"
#include "CColorConverter.h"
#include "SIrrCreationParameters.h"
#include <CoreFoundation/CFRunLoop.h>
#include "SExposedVideoData.h"

//~ #include "irr/base/DeviceIPhone_Apple.h"

namespace irr
{
namespace video
{
	IVideoDriver* createOGLES1Driver(const SIrrlichtCreationParameters& params,
		video::SExposedVideoData& data, io::IFileSystem* io,
		MIrrIPhoneDevice const & device);
}
}

namespace irr
{

//! constructor
CIrrDeviceIPhone::CIrrDeviceIPhone(const SIrrlichtCreationParameters& params)
	: CIrrDeviceStub(params)
	, Close(false)
	, Closed(false)
	, WindowActive(false)
	, WindowMinimized(false)
{
	#ifdef _DEBUG
	setDebugName("CIrrDeviceIPhone");
	#endif
	
	IrrIPhoneDevice.DeviceM = 0;
	IrrIPhoneDevice.DeviceCPP = this;
	IrrIPhoneDevice.displayCreate = 0;
	IrrIPhoneDevice.displayInit = 0;
	IrrIPhoneDevice.displayBegin = 0;
	IrrIPhoneDevice.displayEnd = 0;
	IrrIPhoneDevice.onTerminate = &CIrrDeviceIPhone::onTerminate;
	IrrIPhoneDevice.onWindowActive = &CIrrDeviceIPhone::onWindowActive;
	
	irr_device_iphone_create(&IrrIPhoneDevice);

	// print version, distribution etc.
	struct utsname info;
	uname(&info);
	Operator = new COSOperator((core::stringc("")
			+info.sysname+" "
			+info.release+" "
			+info.version+" "
			+info.machine).c_str());
	os::Printer::log(Operator->getOperationSystemVersion(), ELL_INFORMATION);

	// create display
	if (CreationParams.DriverType != video::EDT_NULL)
	{
		// create the display, only if we do not use the null device
		if (!createDisplay())
			return;
	}

	// create driver
	createDriver();

	if (!VideoDriver)
		return;

	createGUIAndScene();
	
	WindowActive = true;
}



//! destructor
CIrrDeviceIPhone::~CIrrDeviceIPhone()
{
}



bool CIrrDeviceIPhone::createDisplay()
{
	(*IrrIPhoneDevice.displayCreate)(
		&IrrIPhoneDevice,
		&CreationParams.WindowId,
		CreationParams.WindowSize.Width, CreationParams.WindowSize.Height);
	return true;
}


//! create the driver
void CIrrDeviceIPhone::createDriver()
{
	switch(CreationParams.DriverType)
	{
	case video::EDT_OGLES1:
	#ifdef _IRR_COMPILE_WITH_OGLES1_
		{
			video::SExposedVideoData data;
			VideoDriver = video::createOGLES1Driver(
				CreationParams, data, FileSystem, IrrIPhoneDevice);
		}
	#else
		os::Printer::log("No OpenGL-ES1 support compiled in.", ELL_ERROR);
	#endif
		break;

	case video::EDT_SOFTWARE:
	case video::EDT_BURNINGSVIDEO:
	case video::EDT_DIRECT3D8:
	case video::EDT_DIRECT3D9:
	case video::EDT_OPENGL:
		os::Printer::log("This driver is not available. Try OpenGL-ES1 renderer.",
			ELL_ERROR);
		break;

	case video::EDT_NULL:
		VideoDriver = video::createNullDriver(FileSystem, CreationParams.WindowSize);
		break;

	default:
		os::Printer::log("Unable to create video driver of unknown type.", ELL_ERROR);
		break;
	}
}



//! runs the device. Returns false if device wants to be deleted
bool CIrrDeviceIPhone::run()
{
	os::Timer::tick();
	yield();
	Closed = Close;
	return !Close;
}



//! Pause the current process for the minimum time allowed only to allow other processes to execute
void CIrrDeviceIPhone::yield()
{
	struct timespec ts = {0,0};
	nanosleep(&ts, NULL);
}



//! Pause execution and let other processes to run for a specified amount of time.
void CIrrDeviceIPhone::sleep(u32 timeMs, bool pauseTimer=false)
{
	bool wasStopped = Timer ? Timer->isStopped() : true;
	
	struct timespec ts;
	ts.tv_sec = (time_t) (timeMs / 1000);
	ts.tv_nsec = (long) (timeMs % 1000) * 1000000;

	if (pauseTimer && !wasStopped)
		Timer->stop();

	nanosleep(&ts, NULL);

	if (pauseTimer && !wasStopped)
		Timer->start();
}



//! sets the caption of the window
void CIrrDeviceIPhone::setWindowCaption(const wchar_t* text)
{
}



//! presents a surface in the client area
bool CIrrDeviceIPhone::present(video::IImage* image, void * windowId, core::rect<s32>* src)
{
	return false;
}



//! notifies the device that it should close itself
void CIrrDeviceIPhone::closeDevice()
{
	WindowActive = false;
	Close = true;
	CFRunLoopStop(CFRunLoopGetMain());
	while (!Closed) yield();
}



//! returns if window is active. if not, nothing need to be drawn
bool CIrrDeviceIPhone::isWindowActive() const
{
	return WindowActive;
}


bool CIrrDeviceIPhone::isWindowFocused() const
{
	return isWindowActive();
}


bool CIrrDeviceIPhone::isWindowMinimized() const
{
	return !isWindowActive();
}


//! Sets if the window should be resizeable in windowed mode.
void CIrrDeviceIPhone::setResizable(bool resize)
{
}


//! Minimizes the window
void CIrrDeviceIPhone::minimizeWindow()
{
}


//! Maximizes the window if possible.
void CIrrDeviceIPhone::maximizeWindow()
{
}


//! Restore the window to normal size if possible.
void CIrrDeviceIPhone::restoreWindow()
{
}


void CIrrDeviceIPhone::setWindowActive(bool active)
{
	WindowActive = active;
}


void CIrrDeviceIPhone::onTerminate(MIrrIPhoneDevice * dev)
{
	if (dev && dev->DeviceCPP)
	{
		static_cast<irr::CIrrDeviceIPhone*>(dev->DeviceCPP)->closeDevice();
	}
}


void CIrrDeviceIPhone::onWindowActive(MIrrIPhoneDevice * dev, int active)
{
	if (dev && dev->DeviceCPP)
	{
		static_cast<irr::CIrrDeviceIPhone*>(dev->DeviceCPP)->setWindowActive(active != 0);
	}
}


} // end namespace

#endif // _IRR_COMPILE_WITH_IPHONE_DEVICE_

